<!DOCTYPE HTML>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="fr" lang="fr">
  <head>
    <meta charset="utf-8" />

    <meta name="language"  content="fr" />
    <meta name="author"    content="Sylvain Laperche" />
    <meta name="viewport"  content="width=device-width, initial-scale=1.0" />
    <meta name="generator" content="Nanoc 4.12.14" />

    <title>Python : optimisation et couverture de code — grim7reaper</title>

    <link rel="icon"       type="image/png" href="/favicon.png" />
    <link rel="canonical"  type="text/html" href="https://grim7reaper.github.io/blog/2019/12/06/python-optimisation-et-couverture-de-code/" />
    <link rel="license"    type="text/html" href="https://creativecommons.org/licenses/by/4.0/">

    <link rel="stylesheet" type="text/css"  href="/stylesheet.css" />
    <link rel="stylesheet" type="text/css"  href="/syntax.css" />
  </head>
  <body>
    <header id="blog-title">
      <h1><a href="/">grim7reaper</a></h1>
      <p>Un artisan du code</p>
    </header>
    <nav id="navbar">
      <ul>
        <li><a href="/">Accueil</a></li>
        <li><a href="/blog/tags">Catégories</a></li>
        <li><a href="/about">À propos</a></li>
        <li><a href="/atom.xml">Flux Atom</a></li>
      </ul>

      <div id="search-container">
        <form action="https://duckduckgo.com/" method="get">
          <input
            type="hidden"
            name="sites"
            value="grim7reaper.github.io" />
          <input
            class="search"
            type="text"
            name="q"
            placeholder="Recherche…" /><!--
          --><button type="submit">↩</button>
        </form>
      </div>
    </nav>

    

<article class="entry" itemscope itemtype="http://schema.org/BlogPosting">
  <h1>Python : optimisation et couverture de code</h1>

  <p>
  Il est courant de vérifier si nos tests couvrent bien notre code, et en Python
  on utilise généralement
  <a href="https://coverage.readthedocs.io/en/latest/">coverage</a> pour ça.
</p>

<p>
  Cependant, il faut savoir que les optimisations de l’interpréteur peuvent
  interférer, même si elles n’ont pas été activées explicitement via
  l’option <code>-O</code>.
</p>



<h2>Démonstration</h2>

<p>
  Le code suivant est une version simplifiée d’un code plus complexe où j’ai
  rencontré ce problème<sup id="fnref:1"><a href="#fn:1">1</a></sup> :
</p>

    <figure class="code">
      <figcaption>Exemple minimal reproduisant le problème</figcaption>
      <div class="highlight">
        <table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
</pre></td><td class="rouge-code"><pre><span class="c1">#!/usr/bin/env python
# coding: utf-8
</span>
<span class="k">def</span> <span class="nf">fetch</span><span class="p">(</span><span class="n">lst</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lst</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">x</span> <span class="o">&amp;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="nb">Exception</span><span class="p">(</span><span class="s">'error'</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="s">'x is odd'</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">print</span><span class="p">(</span><span class="s">'nothing to do'</span><span class="p">)</span>

<span class="n">fetch</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
</pre></td></tr></tbody></table>
      </div>
    </figure>


<p>
  Si l’on exécute ce code, avec <code>coverage run foo.py</code> on obtient la
  trace suivante :
</p>

    <figure class="code">
      
      <div class="highlight">
        <pre>1
x is odd
nothing to do
2
</pre>
      </div>
    </figure>


<p>On constate que chaque ligne est exécutée au moins une fois :</p>

<ul>
  <li>
    <code>1</code> : <code>print</code> de la ligne 6.
  </li>
  <li>
    <code>x is odd</code> : <code>print</code> de la ligne 11, ce qui signifie
    que l’on est passé par le <code>if</code> de ligne 8 et que l’on a levé une
    exception à la ligne 9.
  </li>
  <li>
    <code>nothing to do</code> : <code>print</code> de la ligne 14.
  </li>
  <li>
    <code>2</code> : <code>print</code> de la ligne 6, étant donné qu’il n’y a
    pas de « nothing to do » on en conclut que l’on est passé par le
    <code>continue</code> de la ligne 13 (et que l’on a sauté
    le <code>print</code> de la ligne 14).
  </li>
</ul>

<p>
  Cependant, si l’on exécute <code>coverage report -m</code> on obtient le
  résultat suivant :
</p>

    <figure class="code">
      
      <div class="highlight">
        <pre>Name     Stmts   Miss  Cover   Missing
--------------------------------------
foo.py      11      1    91%   13
</pre>
      </div>
    </figure>


<p>
  D’après <code>coverage</code>, nous ne sommes pas passé par la ligne 13 !
  Impossible, car si c’était le cas, nous aurions vu « nothing to do » après le
  « 2 ».
</p>

<p>
  Pour aller encore plus loin dans le « fun », on peut modifier le code en
  déplaçant le <code>continue</code> à l’intérieur du <code>try</code>, après la
  levée d’exception :
</p>

    <figure class="code">
      <figcaption>Version alternative</figcaption>
      <div class="highlight">
        <table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
</pre></td><td class="rouge-code"><pre><span class="c1">#!/usr/bin/env python
# coding: utf-8
</span>
<span class="k">def</span> <span class="nf">fetch</span><span class="p">(</span><span class="n">lst</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lst</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">x</span> <span class="o">&amp;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="nb">Exception</span><span class="p">(</span><span class="s">'meh!'</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="s">'x is odd'</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="s">'nothing to do'</span><span class="p">)</span>

<span class="n">fetch</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
</pre></td></tr></tbody></table>
      </div>
    </figure>


<p>
  Nous avons le même comportement, car <code>coverage run foo.py</code> produit
  exactement la même trace :
</p>

    <figure class="code">
      
      <div class="highlight">
        <pre>1
x is odd
nothing to do
2
</pre>
      </div>
    </figure>


<p>
  Par contre, cette fois ci <code>coverage</code> confirme bien que chaque ligne
  a été exécutée :
</p>

    <figure class="code">
      
      <div class="highlight">
        <pre>Name     Stmts   Miss  Cover   Missing
--------------------------------------
ok2.py      11      0   100%
</pre>
      </div>
    </figure>


<p>Étrange n’est-ce pas ?</p>



<h2>Sous le capot</h2>

<p>
  Dans ce genre de cas, il peut être intéressant de voir le <em>bytecode</em>
  généré par l’interpréteur. Pour ce faire, nous allons utiliser le
  module <code>dis</code> (que j’avais déjà
  utilisé <a href="/blog/2013/11/10/crackme-python/">ici</a>) afin de comparer
  le <em>bytecode</em> des deux fonctions et essayer de comprendre l’origine de
  cette différence.
</p>

<h3>Version qui fonctionne</h3>

<p>On va commencer par analyser la version qui fonctionne.</p>

    <figure class="code">
      
      <div class="highlight">
        <pre>  5           0 SETUP_LOOP              72 (to 74)
              2 LOAD_FAST                0 (lst)
              4 GET_ITER
        &gt;&gt;    6 FOR_ITER                64 (to 72)
              8 STORE_FAST               1 (x)

  6          10 LOAD_GLOBAL              0 (print)
             12 LOAD_FAST                1 (x)
             14 CALL_FUNCTION            1
             16 POP_TOP

  7          18 SETUP_EXCEPT            22 (to 42)

  8          20 LOAD_FAST                1 (x)
             22 LOAD_CONST               1 (1)
             24 BINARY_AND
             26 POP_JUMP_IF_FALSE       36

  9          28 LOAD_GLOBAL              1 (Exception)
             30 LOAD_CONST               2 ('meh!')
             32 CALL_FUNCTION            1
             34 RAISE_VARARGS            1

 10     &gt;&gt;   36 CONTINUE_LOOP            6
             38 POP_BLOCK
             40 JUMP_FORWARD            20 (to 62)

 11     &gt;&gt;   42 POP_TOP
             44 POP_TOP
             46 POP_TOP

 12          48 LOAD_GLOBAL              0 (print)
             50 LOAD_CONST               3 ('x is odd')
             52 CALL_FUNCTION            1
             54 POP_TOP
             56 POP_EXCEPT
             58 JUMP_FORWARD             2 (to 62)
             60 END_FINALLY

 13     &gt;&gt;   62 LOAD_GLOBAL              0 (print)
             64 LOAD_CONST               4 ('nothing to do')
             66 CALL_FUNCTION            1
             68 POP_TOP
             70 JUMP_ABSOLUTE            6
        &gt;&gt;   72 POP_BLOCK
        &gt;&gt;   74 LOAD_CONST               0 (None)
             76 RETURN_VALUE
</pre>
      </div>
    </figure>


<p>
  Petite parenthèse sur le format de sortie de <code>dis</code> :
</p>

<ol>
  <li>
    première colonne : numéro de ligne pour la première instruction de chaque
    ligne (une ligne de code Python étant composée de plusieurs instructions).
  </li>
  <li>
    deuxième colonne : <code>--></code> pour indiquer l’instruction en cours
    d’exécution.
  </li>
  <li>
    troisième colonne : <code>>></code> pour indiquer que l’instruction est une
    cible de saut (<em>i.e.</em> une autre instruction peut sauter à cet
    endroit).
  </li>
  <li>
    quatrième colonne : <em>offset</em> de l’instruction (souvent utilisé comme
    cible de saut).
  </li>
  <li>
    cinquième colonne : nom de l’instruction.
  </li>
  <li>
    sixième colonne : arguments de l’instruction.
  </li>
  <li>
    septième colonne : informations supplémentaires sur les arguments.
  </li>
</ol>

<p>
  Sachant cela, on voit bien que notre <code>continue</code> ligne 10 est
  traduit en un <code>CONTINUE_LOOP</code> (et deux autres instructions pour
  sortir du <code>try</code> et retourner au début de la boucle).
</p>

<h3>Version problèmatique</h3>

<p>
  Maintenant, voyons voir le <em>bytecode</em> de la version
  problèmatique :
</p>

    <figure class="code">
      
      <div class="highlight">
        <pre>  5           0 SETUP_LOOP              72 (to 74)
              2 LOAD_FAST                0 (lst)
              4 GET_ITER
        &gt;&gt;    6 FOR_ITER                64 (to 72)
              8 STORE_FAST               1 (x)

  6          10 LOAD_GLOBAL              0 (print)
             12 LOAD_FAST                1 (x)
             14 CALL_FUNCTION            1
             16 POP_TOP

  7          18 SETUP_EXCEPT            20 (to 40)

  8          20 LOAD_FAST                1 (x)
             22 LOAD_CONST               1 (1)
             24 BINARY_AND
             26 POP_JUMP_IF_FALSE       36

  9          28 LOAD_GLOBAL              1 (Exception)
             30 LOAD_CONST               2 ('meh!')
             32 CALL_FUNCTION            1
             34 RAISE_VARARGS            1
        &gt;&gt;   36 POP_BLOCK
             38 JUMP_ABSOLUTE            6

 10     &gt;&gt;   40 POP_TOP
             42 POP_TOP
             44 POP_TOP

 11          46 LOAD_GLOBAL              0 (print)
             48 LOAD_CONST               3 ('x is odd')
             50 CALL_FUNCTION            1
             52 POP_TOP
             54 POP_EXCEPT
             56 JUMP_FORWARD             4 (to 62)
             58 END_FINALLY

 13          60 JUMP_ABSOLUTE            6

 14     &gt;&gt;   62 LOAD_GLOBAL              0 (print)
             64 LOAD_CONST               4 ('nothing to do')
             66 CALL_FUNCTION            1
             68 POP_TOP
             70 JUMP_ABSOLUTE            6
        &gt;&gt;   72 POP_BLOCK
        &gt;&gt;   74 LOAD_CONST               0 (None)
             76 RETURN_VALUE
</pre>
      </div>
    </figure>


<p>
  La première chose que l’on remarque, c’est l’absence
  d’instruction <code>CONTINUE_LOOP</code> : à la place nous avons
  un <code>JUMP_ABSOLUTE</code> à la ligne 13.
</p>

<p>
  La seconde chose que l’on remarque c’est que cette instruction
  <code>JUMP_ABSOLUTE</code> n’est pas atteignable :
</p>

<ul>
  <li>
    si la condition du <code>if</code> est fausse (position 26), on saute à la
    position 36 qui fait un saut à la position 6.
  </li>
  <li>
    si la condition du <code>if</code> est vraie, on va lancer une exception et
    sauter à la position 40 (cf. le <code>SETUP_EXCEPT</code> de la position
    18), puis exécuter les instructions jusqu’au saut de la position 56 qui nous
    amène à la position 62.
  </li>
</ul>

<p>
  Dans les deux cas, on ne passe pas par la position 60, et comme aucune
  instruction ne saute à cette position (elle n’est pas marqué
  avec <code>>></code>) et bien cette instruction n’est jamais exécutée.
</p>



<h3>Hypothèse</h3>

<p>
  Si l’on devait générer le code nous-même en faisant du « mot à mot », on
  produirait probablement la séquence suivante :
</p>

<ul>
  <li>
    on commence par traduire le <code>if</code> de la ligne 8 en utilisant
    un <code>POP_JUMP_IF_FALSE</code> qui saute à la fin du bloc
    <code>try</code> sur un <code>POP_BLOCK</code> (position 36) pour sortir du
    bloc.
  </li>
  <li>
    à la suite de ce <code>POP_BLOCK</code> on ajoute un saut jusqu’au
    <code>continue</code> de la ligne 13/position 60 (probablement en utilisant
    un <code>JUMP_FORWARD</code>).
  </li>
  <li>
    enfin, le <code>continue</code> de la ligne 13 est représenté par
    un <code>CONTINUE_LOOP</code> qui saute au début de la boucle à la position
    6 (comme dans l’exemple qui fonctionne).
  </li>
</ul>

<p>
  Maintenant, si on prend un peu de recul et que l’on décrit les choses de
  manière plus abstraite cela donne :
</p>

<ol>
  <li>
    saute vers A si la condition est fausse (<code>POP_JUMP_IF_FALSE</code> vers
    le <code>POP_BLOCK</code>).
  </li>
  <li>
    saute vers B (<code>JUMP_FORWARD</code> vers le <code>continue</code>).
  </li>
  <li>
    saute vers C (<code>CONTINUE_LOOP</code> vers le début de la boucle).
  </li>
</ol>

<p>
  On remarque que les deux dernières étapes sont des sauts inconditionnels, par
  conséquent l’étape 2 est superflue : on peut sauter de A vers C directement.
</p>

<p>
  On peut donc raisonnablement penser que CPython sait reconnaître ce motif de
  « deux sauts inconditionnels consécutifs » et le remplace par un unique saut
  vers la destination finale.
</p>

<p>
  Ce genre d’optimisations où l’on remplace une suite
  d’instructions <code>A</code> par une autre suite
  d’instructions <code>B</code> fonctionnellement équivalente (mais plus
  efficace) est très commune et s’appelle une
  <a href="https://en.wikipedia.org/wiki/Peephole_optimization">peephole optimization</a>.
</p>

<p>
  Et c’est bien ce que semble faire CPython en sautant directement au début de
  la boucle depuis la fin du bloc <code>try</code>, ignorant totalement notre
  bloc <code>else</code> au passage (qui ne contient rien d’autre qu’un saut
  inconditionnel).
</p>

<p>
  Cela semble être confirmé par le fait que si l’on rajoute
  un <code>print</code> dans le <code>else</code> (avant
  le <code>continue</code>) alors le bloc <code>else</code> et
  son <code>continue</code> sont maintenant exécutés. En effet, dans le cas
  contraire le <code>print</code> ne serait pas exécuté et cela provoquerait une
  différence de comportement observable qui rendrait l’optimisation caduque.
</p>



<h3>Confirmation de l’hypothèse</h3>

<p>
  Bon, on a maintenant une hypothèse plutôt solide, et après quelques recherches
  on tombe sur :
</p>

<ul>
  <li>
    <a href="https://github.com/pytest-dev/pytest-cov/issues/167">ce rapport de bogue</a>
    chez <code>pytest-cov</code> qui nous apprend que la
    bibliothèque <code>coverage</code> serait à l’origine du problème.
  </li>
  <li>
    il s’avère que le lien donné pointe sur
    <a href="https://github.com/nedbat/coveragepy/issues/594">un doublon</a>.
  </li>
  <li>
    en remontant au
    <a href="https://github.com/nedbat/coveragepy/issues/198">bogue d’origine</a>
    (on remarquera le titre qui mentionne <code>continue</code>) on découvre que
    le problème ne vient pas de <code>coverage</code> lui-même mais de
    l’interpréteur Python (CPython en l’occurrence) qui est capable de remplacer
    un saut vers un <code>continue</code> par un saut vers le début de la boucle
    directement (au passage, on y apprend aussi que cette optimisation est
    <strong>toujours</strong> activée).
  </li>
</ul>

<p>
  Pour ceux qui veulent suivre l’évolution de ce problème,
  <a href="https://bugs.python.org/issue2506">un bogue à été ouvert</a> pour
  l’interpréteur.
</p>

<p>
  Cela vient donc confirmer notre hypothèse : l’optimiseur intervient sur les
  instructions générées, ce qui affecte la couverture du code source.
</p>



<h2>Conclusion</h2>

<p>
  La première leçon que l’on peut en tirer c’est que le code que l’on voit ne
  correspond pas forcément au code qui est exécuté. C’est bien connu dans les
  langages tel que le C par exemple, mais c’est plus surprenant en Python.
</p>

<p>
  Deuxième point important : ce problème est simple à résoudre car « il
  suffit de » réécrire le code en utilisant la version « qui fonctionne », et
  c’est même plutôt bénéfique étant donné qu’elle est aussi plus lisible que le
  code d’origine. En effet, l’utilisation du <code>else</code> dans une
  construction <code>try/except</code> est spécifique à Python et donc
  probablement méconnue.
</p>

<p>
  Au passage, ces constructions « bizarres » à base
  de <code>try</code>, <code>else</code> et <code>continue</code>, en plus
  d’être parfois difficile à déchiffrer, ne sont pas toujours bien supportées :
  pendant longtemps il a été impossible d’avoir un
  <code>continue</code> à l’intérieur d’un bloc <code>try</code> (c’est
  maintenant possible). Et aujourd’hui
  encore<sup id="fnref:2"><a href="#fn:2">2</a></sup> il n’est pas possible
  d’avoir un <code>continue</code> dans un bloc <code>finally</code> : c’est une
  erreur de syntaxe !
</p>

<p>
  En conclusion, il vaut mieux écrire du code simple et quand quelque chose de
  bizarre se produit il ne faut pas hésiter à regarder sous le capot pour en
  comprendre les raisons.
</p>



<div class="footnotes">
  <hr class="weak-hr" />
  <ol>
    <li id="fn:1">
      <p>
        le code du <code>else</code> dans la construction
        <code>try</code>/<code>except</code>/<code>else</code> est exécuté
        seulement si aucune exception n’a été levée à l‘intérieur du
        <code>try</code>.<a href="#fnref:1">↩</a>
      </p>
    </li>
    <li id="fn:2">
      <p>
        c’est maintenant possible en
        <a href="https://docs.python.org/3.8/whatsnew/3.8.html#other-language-changes">Python 3.8</a>,
        qui est sorti entre le début de la rédaction de cet article et sa
        publication.<a href="#fnref:2">↩</a>
      </p>
    </li>
  </ol>
</div>


  <hr class="weak-hr" />

  <footer>
    <p>
      <!-- date -->
      Publié le
      <time datetime="2019-12-06T00:00:00Z"
            itemprop="datePublished">
        2019-12-06
      </time>
      
        —
        <!-- tags -->
        Catégories : <a href="/blog/tags/#développement" rel="tag">Développement</a> <a href="/blog/tags/#python" rel="tag">Python</a>
      
    </p>
  </footer>
</article>


    <hr class="strong-hr" />

    <footer>
      <small>
        Copyright © 2010-2023 — Sylvain Laperche
        <br />
        Le contenu de ce site est sous licence
        <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC-BY</a>.
        <br />
        Produit avec <a href="https://nanoc.app/">nanoc</a>.
      </small>
    </footer>
  </body>
</html>
